// @descope/mcp-express/dist/cjs/schemas/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DescopeMcpProviderOptionsSchema = exports.AuthorizationServerOptionsSchema = exports.DynamicClientRegistrationOptionsSchema = exports.VerifyTokenOptionsSchema = exports.PermissionScopeSchema = exports.AttributeScopeSchema = exports.ScopeSchema = void 0;
const zod_1 = require("zod");
/**
 * Base scope definition with common properties.
 */
exports.ScopeSchema = zod_1.z.object({
    /** Name of the scope */
    name: zod_1.z.string(),
    /** Optional description of the scope */
    description: zod_1.z.string().optional(),
    /** Whether this scope is required */
    required: zod_1.z.boolean().optional(),
});
/**
 * Scope definition for attribute access.
 */
exports.AttributeScopeSchema = exports.ScopeSchema.extend({
    /** List of attributes that can be accessed */
    attributes: zod_1.z.array(zod_1.z.string()),
});
/**
 * Scope definition for permission access.
 */
exports.PermissionScopeSchema = exports.ScopeSchema.extend({
    /** List of roles that can be accessed */
    roles: zod_1.z.array(zod_1.z.string()).optional(),
});
/**
 * Options for verifying access tokens.
 */
exports.VerifyTokenOptionsSchema = zod_1.z.object({
    /** Scopes to verify against the token */
    requiredScopes: zod_1.z.array(zod_1.z.string()).optional(),
    /** Audience(s) to verify against the token */
    audience: zod_1.z.array(zod_1.z.string()).optional(),
    /** Key to use for token verification */
    key: zod_1.z.union([zod_1.z.any(), zod_1.z.string()]).optional(),
    /** Resource indicator to validate according to RFC 8707 */
    resourceIndicator: zod_1.z.string().optional(),
});
/**
 * Configuration for dynamic client registration functionality.
 */
exports.DynamicClientRegistrationOptionsSchema = zod_1.z.object({
    /** Whether dynamic client registration is enabled */
    isDisabled: zod_1.z.boolean().optional(),
    /** URL for the authentication flow */
    authPageUrl: zod_1.z.string().optional(),
    /** Scopes for attribute access */
    attributeScopes: zod_1.z.array(exports.AttributeScopeSchema).optional(),
    /** Scopes for permission access */
    permissionScopes: zod_1.z.array(exports.PermissionScopeSchema).optional(),
    /** Logo for the client */
    logo: zod_1.z.string().optional(),
    /** Whether the client is confidential */
    nonConfidentialClient: zod_1.z.boolean().optional(),
});
/**
 * Configuration for OAuth 2.0 Authorization Server functionality.
 */
exports.AuthorizationServerOptionsSchema = zod_1.z.object({
    /** Whether the Authorization Server endpoints are disabled (default: true) */
    isDisabled: zod_1.z.boolean().optional().default(true),
    /** Whether to enable the /authorize endpoint */
    enableAuthorizeEndpoint: zod_1.z.boolean().optional(),
    /** Whether to enable dynamic client registration */
    enableDynamicClientRegistration: zod_1.z.boolean().optional(),
});
/**
 * Configuration options for the Descope MCP SDK.
 */
exports.DescopeMcpProviderOptionsSchema = zod_1.z.object({
    /** The Descope project ID */
    projectId: zod_1.z.string().optional(),
    /** The Descope management key for administrative operations */
    managementKey: zod_1.z.string().optional(),
    /** Descope endpoints are usually hosted on a different subdomain or domain
     * versus the MCP Server itself. This MCP Server URL allows us to
     * handle Dynamic Client Registration on a local server path.
     *  **/
    serverUrl: zod_1.z.string().optional(),
    /** The Descope base URL if a custom domain is set */
    baseUrl: zod_1.z.string().optional(),
    /** Options for dynamic client registration */
    dynamicClientRegistrationOptions: exports.DynamicClientRegistrationOptionsSchema.optional(),
    /** Options for Authorization Server endpoints */
    authorizationServerOptions: exports.AuthorizationServerOptionsSchema.optional(),
    /** Options for token verification */
    verifyTokenOptions: exports.VerifyTokenOptionsSchema.optional(),
    /** Human readable documentation */
    serviceDocumentationUrl: zod_1.z.string().optional(),
});
